### PR TITLE
q40603-proposal-attempt1

### PR DESCRIPTION
**Pull request by q40603**
---
A good PR should include the following parts:

- [ ] A clear title
- [ ] A meaningful message for PR, as well as its commits
- [ ] from your specific branch (***not master***) merging to teacher's `HW` branch (***not master***, again)
- [ ] excluding any irrelevant files, such as binaries, text files, or dot files
- [ ] passing tests/CI

### CODE REVIEW

**Comment by yungyuc** [comment]
Thanks @q40603 .  It seemed that you created the PR from `master` so that it contains unwanted changeset.  You should have created the PR from `proposal`.  To help you remove the unwanted change, I merged the `proposal` branch with `master`.  Could you do either of the following before I start review your proposal?

1. Pull and rebased to the current `proposal` branch.  Update the PR.
2. Use the new `master` / `proposal`, and create another PR.

**Comment by q40603** [comment]
Fix it with rebase.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md:11]: Please add the github link in the proposal.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md]: The template file uses [reStructuredText](https://docutils.sourceforge.io/rst.html), not markdown.  Is the filename a typo?

If you choose markdown, please reformat the file.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md:16]: If you own the images, it's a good idea to add them here in the proposal directory as records.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Now this is markdown.  I am not pretty sure which text format you are trying to use.  It'd be better if you stick with one.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Use English.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Need to add architecture.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Need to add API.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Add engineering.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Add schedule.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Please correct the format and add the missing parts before the proposal due.

**Comment by yungyuc** [comment]
@q40603 When you finish updating your branch please add comments to let me know I should review again.

**Comment by q40603** [comment]
> @q40603 When you finish updating your branch please add comments to let me know I should review again.
Proposal finished.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md:46]: The equations will be good to be included in a written report or presentation slides.  If you have not done it, you are encouraged to do it.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md:147]: This is an excellent question.  It's no and yes.  No, the design patterns for numerical, scientific, and engineering codes are are special.  But yes, the codes do use a special selection and combination of the patterns.

There is not a good reference for the numerical design patterns, but I think the patterns used in game design have good resemblance.  You may take a look at https://gameprogrammingpatterns.com/contents.html .

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md:148]: Your implementation is the starting line of the journey to discover what will be the ways.

File: proposal/q40603/proposal.md
Comment (yungyuc) [inline proposal/q40603/proposal.md:149]: It sounds like a good idea.  I do not see showstopper of using that technique.  Keep in mind that it doesn't solve all problems in the coupling.

**Comment by yungyuc** [comment]
LGTM.  I will merge once I have all proposal graded.

**Review by yungyuc — APPROVED** [review]

**Pull request merged**
